cli/tools/registry/provenance.rs

```diff
@@ -0,0 +1,622 @@
+use std::env;
+
+const PAE_PREFIX: &str = "DSSEv1";
+
+/// DSSE Pre-Auth Encoding
+///
+/// https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#signature-definition
+fn pre_auth_encoding(payload_type: &str, payload: &str) -> Vec<u8> {
+  format!(
+    "{} {} {} {} {}",
+    PAE_PREFIX,
+    payload_type.len(),
+    payload_type,
+    payload.len(),
+    payload,
+  )
+  .into_bytes()
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Signature {
+  keyid: &'static str,
+  sig: String,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Envelope {
+  payload_type: String,
+  payload: String,
+  signatures: Vec<Signature>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct SignatureBundle {
+  #[serde(rename = "$case")]
+  case: &'static str,
+  dsse_envelope: Envelope,
+}
+
+#[derive(Serialize)]
+pub struct SubjectDigest {
+  pub sha512: String,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Subject {
+  pub name: String,
+  pub digest: SubjectDigest,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct GhaResourceDigest {
+  git_commit: String,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct GithubInternalParameters {
+  event_name: String,
+  repository_id: String,
+  repository_owner_id: String,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct ResourceDescriptor {
+  uri: String,
+  digest: Option<GhaResourceDigest>,
+}
+
+#[derive(Serialize)]
+struct InternalParameters {
+  github: GithubInternalParameters,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct GhaWorkflow {
+  #[serde(rename = "ref")]
+  ref_: String,
+  repository: String,
+  path: String,
+}
+
+#[derive(Serialize)]
+struct ExternalParameters {
+  workflow: GhaWorkflow,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct BuildDefinition {
+  build_type: &'static str,
+  resolved_dependencies: [ResourceDescriptor; 1],
+  internal_parameters: InternalParameters,
+  external_parameters: ExternalParameters,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Builder {
+  id: String,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Metadata {
+  invocation_id: String,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct RunDetails {
+  builder: Builder,
+  metadata: Metadata,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Predicate {
+  build_definition: BuildDefinition,
+  run_details: RunDetails,
+}
+
+impl Predicate {
+  pub fn new_github_actions() -> Self {
+    let repo =
+      std::env::var("GITHUB_REPOSITORY").expect("GITHUB_REPOSITORY not set");
+    let rel_ref = std::env::var("GITHUB_WORKFLOW_REF")
+      .unwrap_or_default()
+      .replace(&format!("{}/", &repo), "");
+
+    let delimn = rel_ref.find('@').unwrap();
+    let (workflow_path, workflow_ref) = rel_ref.split_at(delimn);
```

Comment:
```suggestion
    let (workflow_path, mut workflow_ref) = rel_ref.split_at(delimn);
    workflow_ref = &workflow_ref[1..];
```

Such that:

workflow_path = `[0, '@')`
workflow_ref = `('@', len)`